### PR TITLE
⚡ Ignore zero duration timeout

### LINKED ISSUE #2022: Download  receiveTimeout
### Package

dio

### Version

5.3.3

### Operating-System

Android

### Output of `flutter doctor -v`

```shell
┌─────────────────────────────────────────────────────────┐
│ A new version of Flutter is available!                  │
│                                                         │
│ To update to the latest version, run "flutter upgrade". │
└─────────────────────────────────────────────────────────┘
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.13.6, on macOS 14.0 23A344 darwin-arm64, locale
    zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.1)
[✓] VS Code (version 1.77.1)
[✓] Connected device (4 available)
```

### Dart Version

_No response_

### Steps to Reproduce

In previous versions, it was only necessary to set the receiveTimeout of the Download Options to 0 to download without timeout. Now, after changing to the Duration type, the time cannot be unlimited

```dart
Future downloadFile(urlPath, savePath,
      {CancelToken? cancelToken, ProgressCallback? onReceiveProgress}) async {
    Response? response;
    response = await dio.download(urlPath, savePath,
        cancelToken: cancelToken,
        options: Options(receiveTimeout: Duration(seconds: 0)),
        onReceiveProgress: onReceiveProgress ??
            (int count, int total) {
              logI("$count $total");
            });
    return response;
  }
```

### Expected Result

download success

### Actual Result

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: DioException [receive timeout]: The request took longer than 0:00:00.000000 to receive data. It was aborted.

### PR DESCRIPTION
**Pull request by AlexV525**
Resolves #2022.

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### Additional context and info (if any)

I've found that the `IOHttpClientAdapter` uses `Stopwatch` to integrate with `receiveTimeout`, which might cause infinity awaits if the stream has no response forever. This might be the root cause of #1739.
https://github.com/cfug/dio/blob/78f3813a8d8948887198ef628e5a2e2039489b03/dio/lib/src/adapters/io_adapter.dart#L194-L200

### CODE REVIEW

**Comment by kuhnroyal** [comment]
Can we make the timeouts in the options class non-nullable and initialize them with zero? That would make a couple things more readable.

**Comment by AlexV525** [comment]
> Can we make the timeouts in the options class non-nullable and initialize them with zero? That would make a couple things more readable.

You mean zero-timeout is better than non-timeout or sth? I'm okay with both of them though.

**Comment by kuhnroyal** [comment]
Specifically the `RequestOptions.*Timeout` - that would avoid the `options.*Timeout ?? Duration.zero` in all adapters.

**Comment by AlexV525** [comment]
> Specifically the `RequestOptions.*Timeout` - that would avoid the `options.*Timeout ?? Duration.zero` in all adapters.

Ah, then I think we cannot do this because we compose options into `RequestOptions` like `Options.xxx ?? BaseOptions.xxx`. If we change the nullability, the fallback becomes invalid.

**Comment by AlexV525** [comment]
> The `RequestOptions` are used to communicate between `DioMixin` and adapters right? We should be able to handle this in the `DioMixin` in a single place.

I've tried to do this but configs are extended and mixin with multiple classes. It's relatively unactionable at the moment. You can give it a try if you want to.

File: dio/lib/src/adapters/browser_adapter.dart [hunk 1]
```diff
@@ -58,13 +58,11 @@ class BrowserHttpClientAdapter implements HttpClientAdapter {
...
-        receiveTimeout != null &&
-        receiveTimeout > Duration.zero) {
-      xhrTimeout = (connectTimeout + receiveTimeout).inMilliseconds;
+    final sendTimeout = options.sendTimeout ?? Duration.zero;
+    final connectTimeout = options.connectTimeout ?? Duration.zero;
+    final receiveTimeout = options.receiveTimeout ?? Duration.zero;
+    final xhrTimeout = (connectTimeout + receiveTimeout).inMilliseconds;
+    if (xhrTimeout > 0) {
       xhr.timeout = xhrTimeout;
     }
```
Comment (CaiJingLong) [inline dio/lib/src/adapters/browser_adapter.dart:67]: Should we not assume a default value here by browser, give it zero when `xhrTimeout == 0`?

**Review by CaiJingLong — COMMENTED** [review]
No other issues found.

**Pull request merged**